kubestellar/kubestellar
# :sparkles: add control annotation for making auto-downsync optional

## Summary

This PR is for a feature request (https://github.com/kubestellar/kubestellar/issues/1094).

### What's changes
- Introduce an annotation `edge.kubestellar.io/downsync-overwrite: "true" or "false"`. If an object in Workload Description Space (WDS) is set `edge.kubestellar.io/downsync-overwrite: "false"`, Syncer does not overwrite the object in Workload Execution Cluster (WEC).
- Added the test case

### Behavior
When an object is updated,
- if `downsync-overwrite` is true, the downstream object is overwritten by the upstream one (This is nothing difference from the current behavior)
- if `downsync-overwrite` is false, the downstream object is not overwritten

When an object is deleted,
- if `downsync-overwrite` is true, the downstream object is deleted (This is nothing difference from the current behavior)
- if `downsync-overwrite` is false, the downstream object is not deleted

When an object is created,
- if `downsync-overwrite` is true, the downstream object is created (This is nothing difference from the current behavior)
- if `downsync-overwrite` is false, the downstream object is also created

The E2E test case covers above.

### Verification

@clubanderson I pushed the image to my repo:
- `ghcr.io/yana1205/kubestellar/syncer:dev-2023-1008` for main branch
- `ghcr.io/yana1205/kubestellar/syncer:dev-2023-1008-v0.7.0` for release-0.7 branch

Could you verify this change works for the requirement? 

From my end, I verified the following scenario on example1 scenario with `ghcr.io/yana1205/kubestellar/syncer:dev-2023-1008` (not attempt for v0.7).

1. I finished setting up exampl1 scenario. Now I am in WDS for specialstuff.
    ```
    $ kubectl ws .
    Current workspace is "root:my-org:wmw-s".
    ```
1. Make sure that configmap `httpd-htdocs` does not have annotation `edge.kubestellar.io/downsync-overwrite`
    ```
    $ kubectl -n specialstuff get cm httpd-htdocs -o yaml | grep 'edge.kubestellar.io/downsync-overwrite'
    ```
1. `edge.kubestellar.io/downsync-overwrite` is not defined so Syncer works as same as today
    1. Update configmap in WEC (add `aaa: value1`)
        ```
        $ KUBECONFIG=~/.kube/config kubectl --context kind-guilder -n specialstuff edit cm httpd-htdocs     
        apiVersion: v1
        data:
          aaa: value1
          index.html: |
        ...
        ```
    1. The change is overwritten by upstream (WDS) one (namely, `aaa` field will be deleted)
        ```
        $ KUBECONFIG=~/.kube/config kubectl --context kind-guilder -n specialstuff get cm httpd-htdocs -o yaml
        apiVersion: v1
        data:
          index.html: |
        ...
        ```
1. Update annotation of configmap in WDS (add `edge.kubestellar.io/downsync-overwrite: "false"`)
    ```
    $ kubectl -n specialstuff edit cm httpd-htdocs                                                   
    apiVersion: v1
    ...
    kind: ConfigMap
    metadata:
      annotations:
        edge.kubestellar.io/downsync-overwrite: "false"
    ...
    ```
1. From now, the change at WEC will not be overwritten.
    1. Update configmap in WEC (add `aaa: value1`)
        ```
        $ KUBECONFIG=~/.kube/config kubectl --context kind-guilder -n specialstuff edit cm httpd-htdocs     
        apiVersion: v1
        data:
          aaa: value1
          index.html: |
        ...
        ```
    1. The change will not be overwritten by upstream value (namely, `aaa` field will be kept)
        ```
        $ KUBECONFIG=~/.kube/config kubectl --context kind-guilder -n specialstuff get cm httpd-htdocs -o yaml
        apiVersion: v1
        data:
          aaa: value1
          index.html: |
        ```
1. The change from upstream side (WDS) is also not propagated to WEC.
    1. Update configmap in WDS (add `aaa: value2`)
        ```
        $ kubectl -n specialstuff edit cm httpd-htdocs
        apiVersion: v1
        data:
          aaa: value2
          index.html: |
        ```
    1. Syncer never overwrites WEC (namely, `aaa` field will be kept as `value1` in WEC)
        ```
        $ KUBECONFIG=~/.kube/config kubectl --context kind-guilder -n specialstuff get cm httpd-htdocs -o yaml
        apiVersion: v1
        data:
          aaa: value1
          index.html: |
        ```
1. Change downsync-overwrite from false to true (update annotation of the configmap in WDS (update `edge.kubestellar.io/downsync-overwrite: "false"` to `edge.kubestellar.io/downsync-overwrite: "true"`))
    ```
    $ kubectl -n specialstuff edit cm httpd-htdocs                                                   
    apiVersion: v1
    ...
    kind: ConfigMap
    metadata:
      annotations:
        edge.kubestellar.io/downsync-overwrite: "true"
    ...
    ```
1. From now, the Syncer is back to the regular mode that overwrites the configmap in WEC by the upstream one.
    ```
    $ KUBECONFIG=~/.kube/config kubectl --context kind-guilder -n specialstuff get cm httpd-htdocs -o yaml
    apiVersion: v1
    data:
      aaa: value2
      index.html: |
    ```

## Related issue(s)

Fixes #
